Content.Server/Chemistry/EntitySystems/ChemicalReactionSystem.cs

```diff
@@ -0,0 +1,406 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Content.Server.Atmos;
+using Content.Server.Atmos.EntitySystems;
+using Content.Server.Atmos.Components;
+using Content.Server.Atmos;
+using Content.Shared.Administration.Logs;
+using Content.Shared.Chemistry.Components;
+using Content.Shared.Chemistry.Reagent;
+using Content.Shared.Chemistry.Reaction;
+using Content.Shared.Database;
+using Content.Shared.FixedPoint;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Containers;
+
+namespace Content.Server.Chemistry.EntitySystems;
+
+    public sealed class ChemicalReactionSystem : EntitySystem
+    {
+        /// <summary>
+        ///     The maximum number of reactions that may occur when a solution is changed.
+        /// </summary>
+        private const int MaxReactionIterations = 20;
+
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly SharedAudioSystem _audio = default!;
+        [Dependency] private readonly IEntityManager _entityManager = default!;
+        [Dependency] private readonly ISharedAdminLogManager _adminLogger = default!;
+        [Dependency] private readonly AtmosphereSystem _atmosphereSystem = default!;
+
+        /// <summary>
+        ///     A cache of all existant chemical reactions indexed by one of their
+        ///     required reactants.
+        /// </summary>
+    private IDictionary<string, List<ReactionPrototype>> _reactions = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        //Never seen code use SubscribeEvent before
+        //I think this is a better option than SubscribeLocalEvent because it doesnt require a component
+        //Puddles have no components, they could be moved here in the future.
+        _entityManager.EventBus.SubscribeEvent<ReactSolutionEvent>(EventSource.Network, this, OnReactSolutionEvent);
```

Comment:
This is a just a networked event subscription. I.e., `SubscribeNetworkEvent<TEvent>(Handler)`, which isn't what you want.
There are broadcast events which aren't directed at any entity and don't require components, you just use `SubscribeLocalEvent<TEvent>(handler)`.

Though either way, this shouldn't be using an event, just call the method directly.

Reply:
I moved the ``ChemicalReactionSystem`` server side, and the only function that calls it is in ``SolutionContainerSystem``, which is in Content.Shared.
Also ``ChemicalReactionSystem`` needs to be in Content.Server because it needs access to Atmos stuff like ``AtmosphereSystem`` and ``GasMixture``.

Reply:
you can just have an abstract/virtual method in the shared system